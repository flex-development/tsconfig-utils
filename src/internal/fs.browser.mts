/**
 * @file Internal - fs/browser
 * @module tsconfig-utils/internal/fs/browser
 */

import type { FileSystem } from '@flex-development/tsconfig-utils'

/**
 * File system API.
 *
 * @const {FileSystem} fs
 */
const fs: FileSystem = {
  /**
   * Get the contents of a file.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  readFileSync(): never {
    throw new Error('[readFileSync] not implemented')
  },

  /**
   * Read the contents of a directory.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  readdirSync(): never {
    throw new Error('[readdirSync] not implemented')
  },

  /**
   * Get the resolved pathname of a directory or file.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  realpathSync(): never {
    throw new Error('[realpathSync] not implemented')
  },

  /**
   * Get information about a directory or file.
   *
   * @return {never}
   *  Never; not implemented
   * @throws {Error}
   */
  statSync(): never {
    throw new Error('[statSync] not implemented')
  }
}

export default fs
