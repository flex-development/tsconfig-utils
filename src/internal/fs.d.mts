declare module '#internal/fs' {
  import type { FileSystem } from '@flex-development/tsconfig-utils'

  /**
   * File system API.
   *
   * @const {FileSystem} fs
   */
  const fs: FileSystem

  export default fs
}
