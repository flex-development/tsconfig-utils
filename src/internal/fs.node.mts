/**
 * @file Internal - fs/node
 * @module mkbuild/internal/fs/node
 */

import type { FileSystem } from '@flex-development/tsconfig-utils'
import { readdirSync, readFileSync, realpathSync, statSync } from 'node:fs'

/**
 * File system API.
 *
 * @const {FileSystem} fs
 */
const fs: FileSystem = {
  readFile: readFileSync,
  readdir: readdirSync,
  realpath: realpathSync,
  stat: statSync
}

export default fs
