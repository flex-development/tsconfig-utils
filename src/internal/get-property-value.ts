/**
 * @file Internal - getPropertyValue
 * @module tsconfig-utils/internal/getPropertyValue
 */

import type { Nilable, ObjectUnknown } from '@flex-development/tutils'

/**
 * Extracts a property value from the given `object`.
 *
 * Returns `undefined` if the given `property` is not defined.
 *
 * **Note**: If `object` is `null` or `undefined`, it will be converted to an
 * empty object first.
 *
 * @template T - Object type
 * @template K - Object property name
 *
 * @param {Nilable<T>} object - Object to evaluate, `null`, or `undefined`
 * @param {string} property - Object property name
 * @return {T[K] | undefined} Property value or `undefined`
 */
function getPropertyValue<T extends ObjectUnknown, K extends string = string>(
  object: Nilable<T>,
  property: K
): T[K] | undefined {
  return Object.getOwnPropertyDescriptor(object ?? {}, property)?.value as T[K]
}

export default getPropertyValue
