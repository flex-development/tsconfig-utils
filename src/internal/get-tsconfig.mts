/**
 * @file Internal - getTsconfig
 * @module tsconfig-utils/internal/getTsconfig
 */

import isTsconfigHost from '#lib/is-tsconfig-host'
import type { Tsconfig, TsconfigHost } from '@flex-development/tsconfig-utils'
import { shake } from '@flex-development/tutils'

/**
 * Get a {@linkcode Tsconfig} from `target`.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Tsconfig | TsconfigHost | null | undefined} target
 *  Tsconfig or tsconfig host
 * @return {Tsconfig}
 *  TypeScript configuration object
 */
function getTsconfig(
  this: void,
  target: Tsconfig | TsconfigHost | null | undefined
): Tsconfig {
  return shake(isTsconfigHost(target) ? target.tsconfig : target ?? {})
}

export default getTsconfig
