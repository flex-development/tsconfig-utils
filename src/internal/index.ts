/**
 * @file Entry Point - Internal
 * @module tsconfig-utils/internal
 */

export { default as isDirectory } from './is-directory'
export { default as isFile } from './is-file'
export { default as readFile } from './read-file'
export { default as validateFunction } from './validate-function'
export { default as validateString } from './validate-string'
export { default as validateURLString } from './validate-url-string'
