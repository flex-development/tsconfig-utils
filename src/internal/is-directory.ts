/**
 * @file Internal - isDirectory
 * @module tsconfig-utils/internal/isDirectory
 */

import type mlly from '@flex-development/mlly'
import fs from 'node:fs'
import validateURLString from './validate-url-string'

/**
 * Checks if a directory exists at the given module `id`.
 *
 * @param {mlly.ModuleId} id - Module id to evaluate
 * @return {boolean} `true` if directory exists at `id`
 */
const isDirectory = (id: mlly.ModuleId): boolean => {
  try {
    validateURLString(id, 'id')
    return fs.statSync(id, { throwIfNoEntry: false })?.isDirectory() ?? false
  } catch {
    return false
  }
}

export default isDirectory
