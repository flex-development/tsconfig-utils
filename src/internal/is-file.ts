/**
 * @file Internal - isFile
 * @module tsconfig-utils/internal/isFile
 */

import type mlly from '@flex-development/mlly'
import fs from 'node:fs'
import validateURLString from './validate-url-string'

/**
 * Checks if a file exists at the given module `id`.
 *
 * @param {mlly.ModuleId} id - Module id to evaluate
 * @return {boolean} `true` if file exists at `id`
 */
const isFile = (id: mlly.ModuleId): boolean => {
  try {
    validateURLString(id, 'id')
    return fs.statSync(id, { throwIfNoEntry: false })?.isFile() ?? false
  } catch {
    return false
  }
}

export default isFile
