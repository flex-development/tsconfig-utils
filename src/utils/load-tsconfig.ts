/**
 * @file Utilities - loadTsconfig
 * @module tsconfig-utils/utils/loadTsconfig
 */

import type { LoadTsconfigOptions } from '#src/interfaces'
import * as internal from '#src/internal'
import {
  ERR_INVALID_RETURN_VALUE,
  ERR_OPERATION_FAILED,
  type NodeError
} from '@flex-development/errnode'
import * as mlly from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import type {
  CompilerOptions,
  CompilerOptionsValue,
  TSConfig
} from '@flex-development/tsconfig-types'
import {
  isNIL,
  isObjectPlain,
  isString,
  type Nilable,
  type Nullable
} from '@flex-development/tutils'
import { mergeAndCompare } from 'merge-anything'
import sortKeys from 'sort-keys'

/**
 * Reads and parses the [tsconfig][1] file at the given module `id`.
 *
 * If the tsconfig file is found, comments and [byte order marks (BOMs)][2] will
 * be removed before parsing. If successfully parsed, an object representation
 * of the tsconfig file will be returned.
 *
 * [Extending tsconfig files][3] is also supported. If not overwritten, the
 * [`baseUrl`][4], [`outDir`][5], and [`rootDir`][6] properties from the base
 * tsconfig file will be made relative to the tsconfig file being loaded.
 *
 * [1]: https://www.typescriptlang.org/tsconfig
 * [2]: https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8
 * [3]: https://www.typescriptlang.org/tsconfig#extends
 * [4]: https://www.typescriptlang.org/tsconfig#baseUrl
 * [5]: https://www.typescriptlang.org/tsconfig#outDir
 * [6]: https://www.typescriptlang.org/tsconfig#rootDir
 *
 * @param {mlly.ModuleId} id - Module id of tsconfig file
 * @param {LoadTsconfigOptions} [options={}] - Load options
 * @return {Nullable<TSConfig>} `TSConfig` object or `null`
 * @throws {NodeError<Error | TypeError>}
 */
const loadTsconfig = (
  id: mlly.ModuleId,
  options: LoadTsconfigOptions = {}
): Nullable<TSConfig> => {
  const { file = internal.isFile, read = internal.readFile } = options

  // ensure id is an instance of URL or a string
  internal.validateURLString(id, 'id')

  // ensure option schemas
  internal.validateFunction(file, 'options.file')
  internal.validateFunction(read, 'options.read')

  // ensure id is an instance of URL
  id = mlly.toURL(id)

  // ensure module id includes '.json' extension
  if (!id.href.endsWith('.json')) id = mlly.toURL(id.href + '.json')

  // exit early if tsconfig file is not a file
  if (!file(id)) return null

  /**
   * Tsconfig file content.
   *
   * @const {string} content
   */
  const content: string = read(id)

  // exit early if tsconfig file is empty
  if (!content.trim()) return {}

  /**
   * Tsconfig object.
   *
   * @var {Nullable<TSConfig>} tsconfig
   */
  let tsconfig: Nullable<TSConfig> = {}

  // parse tsconfig file
  try {
    tsconfig = internal.parseJSON<TSConfig>(content)
  } catch (e: unknown) {
    throw new ERR_OPERATION_FAILED((e as Error).message)
  }

  // throw if tsconfig is not plain object
  if (!isObjectPlain(tsconfig)) {
    throw new ERR_INVALID_RETURN_VALUE('a plain object', 'parseJSON', tsconfig)
  }

  /**
   * Configuration files to inherit from.
   *
   * @const {string[]} bases
   */
  const bases: string[] = isNIL(tsconfig.extends)
    ? []
    : [tsconfig.extends].flat().filter(extend => !!extend.trim())

  // try merging tsconfig and base tsconfigs
  for (const extend of bases) {
    /**
     * Absolute path to base tsconfig file.
     *
     * @const {string} basepath
     */
    const basepath: string = pathe.join(
      pathe.dirname(id.pathname),
      (extend + '.json').replace(/(\.json\.json)$/, '.json')
    )

    /**
     * Base tsconfig object.
     *
     * @const {Nullable<TSConfig>} base
     */
    const base: Nullable<TSConfig> = loadTsconfig(basepath, { file, read })

    // merge tsconfig objects if base tsconfig object was found
    if (base) {
      /**
       * Compares values from a base tsconfig file and inheriting tsconfig
       * file to determine how the values should be merged.
       *
       * @param {Nilable<CompilerOptions | CompilerOptionsValue>} b - Value
       * from base tsconfig object, {@linkcode base}
       * @param {Nilable<CompilerOptions | CompilerOptionsValue>} t - Value
       * from inheriting tsconfig object, {@linkcode tsconfig}
       * @param {string | symbol} key - Object key being evaluated
       * @return {Nilable<CompilerOptions | CompilerOptionsValue>} Merge value
       */
      const compare = (
        b: Nilable<CompilerOptions | CompilerOptionsValue>,
        t: Nilable<CompilerOptions | CompilerOptionsValue>,
        key: string | symbol
      ): Nilable<CompilerOptions | CompilerOptionsValue> => {
        /**
         * Merge value.
         *
         * @var {Nilable<CompilerOptions | CompilerOptionsValue>} merged
         */
        let merged: Nilable<CompilerOptions | CompilerOptionsValue> = t

        // determine how to merge values
        switch (true) {
          // relative paths should be interpreted as relative to base, but
          // they need also need to be relative to inheriting tsconfig
          case key === 'baseUrl' && isString(b):
          case key === 'outDir' && isString(b):
          case key === 'rootDir' && isString(b):
            if (b === t) merged = pathe.join(pathe.dirname(extend), b as string)
            break
          // recursively merge compilerOptions
          case key === 'compilerOptions':
            merged = mergeAndCompare(compare, b, t) as CompilerOptions
            merged = sortKeys(merged)
            break
          // exclude, files, and include properties from inheriting tsconfig
          // file should overwrite those from base tsconfig file.
          case key === 'exclude':
          case key === 'files':
          case key === 'include':
            merged = t ?? /* c8 ignore next */ b
            break
          // references is the only top-level property excluded from inheritance
          case key === 'references':
            break
          default:
            break
        }

        return merged
      }

      // merge tsconfig objects
      tsconfig = mergeAndCompare(compare, base, tsconfig)
    }
  }

  return sortKeys(tsconfig)
}

export default loadTsconfig
