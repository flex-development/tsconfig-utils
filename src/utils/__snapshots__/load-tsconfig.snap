// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/loadTsconfig > should return TSConfig object if tsconfig file is found 1`] = `
{
  "compilerOptions": {
    "allowImportingTsExtensions": true,
    "allowJs": true,
    "allowUnreachableCode": false,
    "alwaysStrict": false,
    "baseUrl": "..",
    "checkJs": false,
    "customConditions": [
      "import",
      "node",
      "default",
    ],
    "declaration": false,
    "declarationMap": false,
    "emitDecoratorMetadata": false,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "experimentalDecorators": false,
    "forceConsistentCasingInFileNames": true,
    "lib": [
      "es2020",
    ],
    "module": "esnext",
    "moduleResolution": "bundler",
    "newLine": "lf",
    "noEmit": true,
    "noEmitOnError": true,
    "noErrorTruncation": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "outDir": "../dist",
    "paths": {
      "#fixtures/*": [
        "__fixtures__/*",
      ],
      "#src": [
        "src/index",
      ],
      "#src/*": [
        "src/*",
      ],
      "#tests/*": [
        "__tests__/*",
      ],
      "@flex-development/tsconfig-utils": [
        "src/index",
      ],
    },
    "preserveConstEnums": true,
    "preserveSymlinks": false,
    "pretty": true,
    "resolveJsonModule": true,
    "resolvePackageJsonExports": true,
    "resolvePackageJsonImports": true,
    "rootDir": "..",
    "skipDefaultLibCheck": false,
    "skipLibCheck": false,
    "sourceMap": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "target": "es2020",
    "useDefineForClassFields": true,
    "useUnknownInCatchVariables": true,
    "verbatimModuleSyntax": true,
  },
  "exclude": [
    "**/__mocks__/**",
    "**/__tests__/**",
  ],
  "extends": "../tsconfig",
  "include": [
    "../dist/*",
    "../src/*",
  ],
  "references": [],
}
`;
